File: java-sample/src/main/java/com/moengage/sample/java/HomeListAdapter.java
package com.moengage.sample.java;

import android.view.LayoutInflater;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.moengage.sample.java.databinding.FeatureListItemBinding;
import java.util.List;

/**
 * @author Arshiya Khanum
 */
public class HomeListAdapter extends RecyclerView.Adapter<HomeListAdapter.ViewHolder> {

  private final List<HomeCategory> optionsList;
  private final OnHomeCategorySelected onHomeCategorySelected;

  public HomeListAdapter(List<HomeCategory> optionsList, OnHomeCategorySelected onHomeCategorySelected) {
    this.optionsList = optionsList;
    this.onHomeCategorySelected = onHomeCategorySelected;
  }

  @NonNull @Override
  public HomeListAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
    return new ViewHolder(
        FeatureListItemBinding.inflate(LayoutInflater.from(parent.getContext()), parent,
        false));
  }

  @Override public void onBindViewHolder(@NonNull HomeListAdapter.ViewHolder holder, int position) {
    holder.binding.featureTitle.setText(optionsList.get(position).getLabel());
    holder.binding.getRoot().setOnClickListener(
        v -> onHomeCategorySelected.onHomeCategorySelected(optionsList.get(position).getCategory()));
  }

  @Override public int getItemCount() {
    return optionsList.size();
  }

  public class ViewHolder extends RecyclerView.ViewHolder {

    public FeatureListItemBinding binding;

    public ViewHolder(@NonNull FeatureListItemBinding binding) {
      super(binding.getRoot());
      this.binding = binding;
    }
  }

  interface OnHomeCategorySelected {

    public void onHomeCategorySelected(FeaturesCategory category);

  }
}

File: kotlin-sample/src/main/java/com/moengage/sample/kotlin/HomeListAdapter.kt
package com.moengage.sample.kotlin

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.moengage.sample.kotlin.databinding.FeatureListItemBinding

/**
 * @author Arshiya Khanum
 */
class HomeListAdapter(
    private val optionsList: List<HomeCategory>,
    private val categorySelected: OnHomeCategorySelected
) : RecyclerView.Adapter<HomeListAdapter.ViewHolder>() {


    override fun onBindViewHolder(viewHolder: ViewHolder, position: Int) {
        viewHolder.binding.featureTitle.text = optionsList[position].label
        viewHolder.binding.root.setOnClickListener {
            categorySelected.onHomeCategorySelected(optionsList[position].category)
        }
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, position: Int): ViewHolder {
        return ViewHolder(
            FeatureListItemBinding.inflate(
                LayoutInflater.from(viewGroup.context),
                viewGroup, false
            )
        )
    }

    override fun getItemCount(): Int {
        return optionsList.size
    }

    class ViewHolder(val binding: FeatureListItemBinding) : RecyclerView.ViewHolder(binding.root)
}

interface OnHomeCategorySelected {
    fun onHomeCategorySelected(category: FeaturesCategory)
}
